pool/context_pool.go

```diff
@@ -23,14 +23,28 @@ type ContextPool struct {
 // are busy, a call to Go() will block until the task can be started.
 func (p *ContextPool) Go(f func(ctx context.Context) error) {
 	p.errorPool.Go(func() error {
+		if p.cancelOnError {
+			// If we are cancelling on error, then we also want to cancel if a
+			// panic is raised. To do this, we need to recover, cancel, and then
+			// re-throw the caught panic.
+			defer func() {
+				if r := recover(); r != nil {
+					p.cancel()
+					panic(r)
+				}
+			}()
+		}
+
 		err := f(p.ctx)
-		if err != nil && p.cancelOnError {
+		if err != nil {
 			// Leaky abstraction warning: We add the error directly because
 			// otherwise, canceling could cause another goroutine to exit and
 			// return an error before this error was added, which breaks the
 			// expectations of WithFirstError().
-			p.errorPool.addErr(err)
-			p.cancel()
+			if err != nil {
+				p.errorPool.addErr(err)
+				p.cancel()
+			}
```

Comment:
why are we checking `if err != nil` in this block again?

Reply:
fixed!